### ISSUE TITLE
System.InvalidCastException during upload

### ISSUE THREAD

**Issue by ogz** [tusdotnet/tusdotnet#198]
Hi,

We are receiving random exceptions in the backend as follows. 
It does not happen always, something around %5 of uploads.
```

Connection ID "xxx", Request ID "yyy": An unhandled exception was thrown by the application.
[2023-01-24 10:17:43.6217 ERROR Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer]
System.InvalidCastException: Unable to cast object of type 'System.Byte[]' to type 'System.IO.Pipelines.BufferSegment'.
   at System.IO.Pipelines.StreamPipeReader.AdvanceTo(SequencePosition consumed, SequencePosition examined)
   at System.IO.Pipelines.StreamPipeReader.AdvanceTo(SequencePosition consumed)
   at tusdotnet.Models.ClientDisconnectGuardedPipeReader.<>c__DisplayClass4_0.<AdvanceTo>b__0()
   at tusdotnet.Helpers.ClientDisconnectGuard.Execute(Action guardFromClientDisconnect, CancellationToken requestCancellationToken)
   at tusdotnet.Models.ClientDisconnectGuardedPipeReader.AdvanceTo(SequencePosition consumed)
   at tusdotnet.Stores.TusDiskStore.AppendDataAsync(String fileId, PipeReader reader, CancellationToken cancellationToken)
   at tusdotnet.IntentHandlers.WriteFileHandler.HandlePipelineWrite()
   at tusdotnet.IntentHandlers.WriteFileHandler.Invoke()
   at tusdotnet.TusProtocolHandlerIntentBased.Invoke(ContextAdapter context)
   at tusdotnet.TusCoreMiddleware.Invoke(HttpContext context)
   at PPMRoom.API.Startup.<>c.<<Configure>b__9_1>d.MoveNext() in E:\GitLab-Runner\builds\NnDCcrQP\0\Api\Startup.cs:line 162
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()
===========
```

Startup.cs:line 162 is where we initialize TUS
`app.UseTus(httpContext => new DefaultTusConfiguration`

**Platform Configuration**
Tusdotnet -> 2.5.0
dotnet -> 6.0

**Comment by smatsson** [comment]
That's an odd issue. `System.IO.Pipelines.StreamPipeReader.Advance` is a method built into .NET and it seems to be doing a cast on this line https://source.dot.net/#System.IO.Pipelines/System/IO/Pipelines/StreamPipeReader.cs,90

I really have no idea why it would have a byte[] instead of a BufferSegment as tusdotnet only forwards the `SequencePosition` received from the earlier ReadAsync call. The SequencePosition is never created from tusdotnet but is all handled by .NET.

Are you running the latest patch version of the .NET runtime on your server?

**Comment by smatsson** [comment]
After some more investigation I found something that could be the culprit. The good news is however that it's not a dangerous exception but something that happens after the client has disconnected.

By the look of the stacktrace you are running some kind of IIS integration. Would you mind posting your Startup/Program class, especially the part where you setup the AppBuilder? `var app = WebApplication.CreateBuilder` and onwards :)

**Comment by smatsson** [comment]
[This commit](https://github.com/tusdotnet/tusdotnet/commit/d08a145bcbb7fb8ff1bac60464606b2093afc9ae) should fix this issue.

**Comment by smatsson** [comment]
Fixed in 2.7.1
